README.md
# flutter_sound_suite

a flutter sound suite for record and play voice easily!

## dependencies:

### dio: ^4.0.0

### intl: ^0.17.0

### uuid: ^3.0.4

### provider: ^5.0.0

### path_provider: ^2.0.2

### flutter_sound: ^8.1.4

#### Actually, the following codecs are supported by flutter_sound:

| | AAC ADS | Opus OGG | Opus CAF | MP3 | Vorbis OGG | PCM raw| PCM WAV | PCM AIFF | PCM CAF | FLAC | AAC MP4 | AMR-NB |
AMR-WB | | :-------------- | :-----: | :------: | :------: | :-- | :--------- | :-- | :-----: | :------: | :-----: | :
--: | :-----: | :----: | :----: | | iOS encoder | Yes | Yes(*) | Yes | No | No | No | Yes | No | Yes | Yes | Yes | NO |
NO | | iOS decoder | Yes | Yes(*) | Yes | Yes | No | No | Yes | Yes | Yes | Yes | Yes | NO | NO | | Android encoder |
Yes | No | No | No | No | Yes | Yes | No | No | No | No | YES | YES | | Android decoder | Yes | Yes | Yes(*)   | Yes |
Yes | Yes | Yes | Yes(*)   | Yes(*)  | Yes | Yes | YES | YES |

This table will eventually be upgrated when more codecs will be added.

Yes(*) : The codec is supported by Flutter Sound, but with a File Format Conversion. This has several drawbacks :

- Needs FFmpeg. FFmpeg is not included in the LITE flavor of Flutter Sound
- Can add some delay before Playing Back the file, or after stopping the recording. This delay can be substancial for
  very large records.

-------------------------------------------------------------------------------------------------------------------------------------

### permission_handler: ^8.0.1

CHANGELOG.md
## [2.1.0] 2021.6.8
* upgrade flutter sdk and plugins

## [2.0.0] 2021.6.8
* upgrade flutter sdk and plugins

## [1.0.0] 2021.3.17
* reformat code

## [0.2.0+2] 2021.3.15

* upgrade flutter sound version to 8.0.0+6
* upgrade dio version to 4.0.0-beta7

## [0.2.0] 2021.3.10

* upgrade flutter sdk version to 2.0.1
* upgrade flutter sound version to 8.0.0+2
* null safety support

## [0.1.0+2] 2020.12.10

* upgrade flutter sound version to 6.4.7+1

## [0.1.0+1] 2020.11.30

* remove compute duration for current playing uri

## [0.1.0] 2020.11.27

* use color: Colors.transparent instead of type: MaterialType.transparency,and fill full screen
* minSdkVersion 24

## [0.0.1+3] 2020.8.29

* unlock flutter_sound's version,but now it only works on version 5.0.1
* 解锁了flutter_sound的版本，不过目前只能在5.0.1上才能用呀

## [0.0.1+2] 2020.8.28

* rename example project

## [0.0.1+1] 2020.8.28

* flutter_sound's version downgrade and locked to 5.0.1,Because 5.1.1 is not available on iOS
* flutter_sound的版本被降级并锁定在5.0.1，因为5.1.1在iOS不可用

## [0.0.1] 2020.8.16

* initial release.

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_sound_suite/flutter_sound_suite.dart';
import 'package:provider/provider.dart';

void main() {
  runApp(FlutterSoundSuiteApp());
}

class FlutterSoundSuiteApp extends StatefulWidget {
  @override
  _FlutterSoundSuiteAppState createState() => _FlutterSoundSuiteAppState();
}

class _FlutterSoundSuiteAppState extends State<FlutterSoundSuiteApp> {
  List<Map> data = [];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Sound Suite Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Flutter Sound Suite Demo'),
        ),
        body: ChangeNotifierProvider(
          create: (context) => SoundPlayerModel(),
          child: Container(
            child: ListView.builder(
              itemBuilder: (BuildContext context, int index) {
                Map item = data[index];
                return Card(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(item['fileName']),
                      Text('duration ${item['duration']}'),
                      Card(
                        color: Colors.red.shade50,
                        child: SoundPlayerWidget(
                          filePath: item['filePath'],
                          fileName: item['fileName'],
                          duration: item['duration'],
                        ),
                      ),
                    ],
                  ),
                );
              },
              itemCount: data.length,
            ),
          ),
        ),
        bottomNavigationBar: SafeArea(
          child: Container(
            height: 70,
            child: SoundRecorderWidget(recordComplete:
                (String filePath, String fileName, double duration) {
              setState(() {
                data.add({
                  'filePath': filePath,
                  'fileName': fileName,
                  'duration': duration,
                });
                debugPrint('save to $filePath duration $duration');
              });
            }),
          ),
        ),
      ),
    );
  }
}


lib/src/sound_player_model.dart
import 'dart:async';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter_sound/flutter_sound.dart';
import 'package:path_provider/path_provider.dart';

class SoundPlayerModel with ChangeNotifier {
  FlutterSoundPlayer _player = FlutterSoundPlayer();
  StreamSubscription? _playerSubscription;
  double _currentPlayTime = 0.0;

  double get currentPlayTime => _currentPlayTime;

  ///maybe remote or local file path
  String? _currentPlayingPath;

  Dio dio = Dio(BaseOptions(
    connectTimeout: const Duration(seconds: 10).inMilliseconds,
    receiveTimeout: const Duration(minutes: 60).inMilliseconds,
  ));
  bool _isDownloading = false;

  bool get isDownloading => _isDownloading;

  SoundPlayerModel() {
    init();
  }

  /// init player
  init() async {
    await _player.closeAudioSession();
    await _player.openAudioSession(
      focus: AudioFocus.requestFocusTransient,
      category: SessionCategory.playAndRecord,
      mode: SessionMode.modeDefault,
    );
    await _player.setSubscriptionDuration(Duration(milliseconds: 10));
  }

  /// release resource here
  @override
  dispose() {
    super.dispose();
    _releasePlayerSubscriptions();
    _releasePlayer();
    dio.close(force: true);
  }

  /// determine given path is current playing file or not
  bool isCurrentPlaying({required String filePath}) =>
      _player.isPlaying && _currentPlayingPath == filePath;

  /// determine player stopped or not
  bool isPlayerStopped() => _player.isStopped;

  /// start of stop player
  startOrStopPlayer(
      {required String filePath,
      required String fileName,
      required Codec codec}) async {
    _currentPlayingPath = filePath;
    Directory tempDir = await getTemporaryDirectory();
    String localPath = '${tempDir.path}/$fileName';
    File localFile = File(localPath);
    debugPrint(
        'startOrStopPlayer $_currentPlayingPath local_exists ${localFile.existsSync()}');
    if (!localFile.existsSync()) {
      _isDownloading = true;
      notifyListeners();
      await dio.download(_currentPlayingPath!, localPath);
      _isDownloading = false;
      notifyListeners();
    }
    if (localFile.existsSync()) {
      debugPrint(
          'startOrStopPlayer $_currentPlayingPath download to $localPath');
      if (_player.isStopped) {
        startPlayer(path: localPath, codec: codec);
      } else {
        stopPlayer();
      }
    }
  }

  _releasePlayerSubscriptions() {
    _playerSubscription?.cancel();
    _playerSubscription = null;
  }

  _releasePlayer() async {
    try {
      await _player.closeAudioSession();
    } catch (e) {
      debugPrint('Released unsuccessful');
      debugPrint(e.toString());
    }
  }

  _addListeners() {
    _releasePlayerSubscriptions();
    _playerSubscription = _player.onProgress!.listen((event) {
      _currentPlayTime = event.position.inMilliseconds / 1000.0;
      notifyListeners();
    });
  }

  /// start player
  startPlayer({required String path, Codec codec: Codec.amrNB}) async {
    try {
      await _player.startPlayer(
          fromURI: path,
          codec: codec,
          whenFinished: () {
            debugPrint('Play finished');
            notifyListeners();
          });
      _addListeners();
      debugPrint('startPlayer');
    } catch (err) {
      debugPrint('error: $err');
    }
    notifyListeners();
  }

  /// stop player
  stopPlayer() async {
    try {
      await _player.stopPlayer();
      debugPrint('stopPlayer');
      _releasePlayerSubscriptions();
    } catch (err) {
      debugPrint('error: $err');
    }
    notifyListeners();
  }

  /// pause or resume player
  pauseResumePlayer() async {
    if (_player.isPlaying) {
      await _player.pausePlayer();
    } else {
      await _player.resumePlayer();
    }
    notifyListeners();
  }

  /// seek to player by milliSeconds
  seekToPlayer(int milliSeconds) async =>
      await _player.seekToPlayer(Duration(milliseconds: milliSeconds));
}


lib/src/sound_player_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter_sound/flutter_sound.dart';
import 'package:provider/provider.dart';

import 'sound_player_model.dart';

class SoundPlayerWidget extends StatefulWidget {
  final String filePath;
  final String fileName;
  final double duration;
  final Codec codec;

  SoundPlayerWidget({
    required this.filePath,
    required this.fileName,
    required this.duration,
    this.codec: Codec.amrNB,
  });

  @override
  _SoundPlayerWidgetState createState() => _SoundPlayerWidgetState();
}

class _SoundPlayerWidgetState extends State<SoundPlayerWidget> {
  @override
  Widget build(BuildContext context) {
    return Container(
      height: 35,
      width: 80 + widget.duration * 10.0,
      child: InkWell(
        borderRadius: BorderRadius.circular(4),
        onTap: () => Provider.of<SoundPlayerModel>(context, listen: false)
            .startOrStopPlayer(
          filePath: widget.filePath,
          fileName: widget.fileName,
          codec: widget.codec,
        ),
        child: Stack(
          alignment: Alignment.centerLeft,
          children: [
            Visibility(
              visible: Provider.of<SoundPlayerModel>(context)
                  .isCurrentPlaying(filePath: widget.filePath),
              child: ClipRRect(
                borderRadius: BorderRadius.circular(4),
                child: LinearProgressIndicator(
                  minHeight: 35,
                  value:
                      Provider.of<SoundPlayerModel>(context).currentPlayTime /
                          widget.duration,
                  valueColor:
                      AlwaysStoppedAnimation<Color>(Colors.orange.shade300),
                  backgroundColor: Colors.white,
                ),
              ),
            ),
            Row(
              children: [
                SizedBox(
                  width: 6,
                ),
                Icon(Provider.of<SoundPlayerModel>(context)
                        .isCurrentPlaying(filePath: widget.filePath)
                    ? Icons.stop
                    : Icons.play_arrow),
                SizedBox(
                  width: 6,
                ),
                Text('${widget.duration}s'),
              ],
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/sound_recorder_widget.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_sound/flutter_sound.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:intl/intl.dart' show DateFormat;
import 'package:path_provider/path_provider.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:uuid/uuid.dart';

typedef RecordCompleteCallback = Function(
    String filePath, String fileName, double duration);

///see[Codec]       AAC_ADS Opus_OGG  Opus_CAF  MP3  Vorbis_OGG  PCM_raw  PCM_WAV  PCM_AIFF  PCM_CAF  FLAC  AAC_MP4 AMR-NB  AMR-WB
///iOS encoder      Yes     Yes(*)    Yes       No   No          No       Yes      No        Yes      Yes   Yes     NO      NO
///iOS decoder      Yes     Yes(*)    Yes       Yes  No          No       Yes      Yes       Yes      Yes   Yes     NO      NO
///Android encoder  Yes     No        No        No   No          Yes      Yes      No        No       No    No      YES     YES
///Android decoder  Yes     Yes       Yes(*)    Yes  Yes         Yes      Yes      Yes(*)    Yes(*)   Yes   Yes     YES     YES

class SoundRecorderWidget extends StatefulWidget {
  final RecordCompleteCallback recordComplete;
  final Codec codec;

  SoundRecorderWidget({required this.recordComplete, this.codec: Codec.amrNB});

  @override
  _SoundRecorderWidgetState createState() => new _SoundRecorderWidgetState();
}

class _SoundRecorderWidgetState extends State<SoundRecorderWidget> {
  double _startY = 0.0;
  double _offset = 0.0;
  bool _isCancel = false;
  String _buttonHint = '按住 说话';
  String _overLayHint = '上滑 取消';
  String _voiceImage = 'assets/voice/volume_1.png';
  OverlayEntry? _overlayEntry;
  String _recordTime = '00:00:00';
  bool _isRecording = false;

  FlutterSoundRecorder _recorder = FlutterSoundRecorder();
  StreamSubscription? _recorderSubscription;

  Codec _codec = Codec.amrNB;
  late String _currentFilePath;
  late String _currentFileName;

  init() async {
    _codec = widget.codec;
    await _recorder.openAudioSession(
      focus: AudioFocus.requestFocusTransient,
      category: SessionCategory.playAndRecord,
      mode: SessionMode.modeDefault,
    );
    await _recorder.setSubscriptionDuration(Duration(milliseconds: 10));
    initializeDateFormatting();
    PermissionStatus status = await Permission.microphone.request();
    if (status != PermissionStatus.granted) {
      throw RecordingPermissionException("Microphone permission not granted");
    }
  }

  @override
  initState() {
    super.initState();
    init();
  }

  @override
  dispose() {
    releaseRecorderSubscriptions();
    releaseRecorder();
    super.dispose();
  }

  releaseRecorderSubscriptions() {
    _recorderSubscription?.cancel();
    _recorderSubscription = null;
  }

  releaseRecorder() async {
    try {
      await _recorder.closeAudioSession();
    } catch (e) {
      debugPrint('Released unsuccessful');
      debugPrint(e.toString());
    }
  }

  startRecorder() async {
    try {
      PermissionStatus status = await Permission.microphone.request();
      if (status != PermissionStatus.granted) {
        throw RecordingPermissionException("Microphone permission not granted");
      }

      Directory tempDir = await getTemporaryDirectory();
      _currentFileName = '${Uuid().v4()}${ext[_codec.index]}';
      _currentFilePath = '${tempDir.path}/$_currentFileName';
      await _recorder.startRecorder(
        toFile: _currentFilePath,
        codec: _codec,
      );
      debugPrint('startRecorder');

      _recorderSubscription = _recorder.onProgress!.listen((event) {
        DateTime date = new DateTime.fromMillisecondsSinceEpoch(
            event.duration.inMilliseconds,
            isUtc: true);
        String txt = DateFormat('mm:ss:SS', 'en_GB').format(date);
        this.setState(() {
          _recordTime = txt.substring(0, 8);
          double _decibels = event.decibels!;
          if (_decibels > 0.0 && _decibels <= 10.0) {
            _voiceImage = 'assets/voice/volume_1.png';
          } else if (_decibels > 10.0 && _decibels <= 20.0) {
            _voiceImage = 'assets/voice/volume_2.png';
          } else if (_decibels > 20.0 && _decibels <= 30.0) {
            _voiceImage = 'assets/voice/volume_3.png';
          } else if (_decibels > 30.0 && _decibels <= 40.0) {
            _voiceImage = 'assets/voice/volume_4.png';
          } else if (_decibels > 40.0 && _decibels <= 50.0) {
            _voiceImage = 'assets/voice/volume_5.png';
          } else if (_decibels > 50.0 && _decibels <= 60.0) {
            _voiceImage = 'assets/voice/volume_6.png';
          } else if (_decibels > 60.0 && _decibels <= 70.0) {
            _voiceImage = 'assets/voice/volume_7.png';
          } else if (_decibels > 70.0) {
            _voiceImage = 'assets/voice/volume_7.png';
          } else {
            _voiceImage = 'assets/voice/volume_1.png';
          }
          _overlayEntry?.markNeedsBuild();
        });
      });

      this.setState(() {
        this._isRecording = true;
      });
    } catch (err) {
      debugPrint('startRecorder error: $err');
      setState(() {
        stopRecorder();
        releaseRecorderSubscriptions();
      });
    }
  }

  stopRecorder() async {
    try {
      await _recorder.stopRecorder();
      debugPrint('stopRecorder');
      releaseRecorderSubscriptions();
      Duration? duration = await flutterSoundHelper.duration(_currentFilePath);
      if (!_isCancel) {
        widget.recordComplete.call(_currentFilePath, _currentFileName,
            duration!.inMilliseconds / 1000.0);
      }
    } catch (err) {
      debugPrint('stopRecorder error: $err');
    }
    this.setState(() {
      this._isRecording = false;
    });
  }

  startStopRecorder() {
    if (_recorder.isRecording || _recorder.isPaused) {
      stopRecorder();
    } else {
      startRecorder();
    }
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onLongPressStart: (details) {
        _startY = details.globalPosition.dy;
        showOverlay();
      },
      onLongPressEnd: (details) {
        hideOverlay();
      },
      onLongPressMoveUpdate: (details) {
        _offset = details.globalPosition.dy;
        updateStatus();
      },
      child: Container(
        width: double.infinity,
        margin: const EdgeInsets.only(top: 7, bottom: 7),
        decoration: BoxDecoration(
          color: Colors.white,
          borderRadius: BorderRadius.circular(4),
        ),
        child: Center(
          child: Text(
            '$_buttonHint${_isRecording ? _recordTime : ''}',
            style: TextStyle(
              fontSize: 18.0,
              color: Colors.black87,
            ),
          ),
        ),
      ),
    );
  }

  showOverlay() {
    setState(() {
      _buttonHint = '松开 结束';
    });
    insertOverLay(context);
    startRecorder();
  }

  hideOverlay() {
    setState(() {
      _buttonHint = '按住 说话';
    });

    stopRecorder();

    _overlayEntry?.remove();
    _overlayEntry = null;
  }

  insertOverLay(BuildContext context) {
    if (_overlayEntry == null) {
      _overlayEntry = OverlayEntry(builder: (content) {
        return Positioned.fill(
          child: Material(
            color: Colors.transparent,
            child: Center(
              child: Opacity(
                opacity: 0.8,
                child: Container(
                  width: 160,
                  height: 160,
                  decoration: BoxDecoration(
                    color: Color(0xff77797A),
                    borderRadius: BorderRadius.all(Radius.circular(20.0)),
                  ),
                  child: Column(
                    children: <Widget>[
                      Container(
                        margin: EdgeInsets.only(top: 10),
                        child: Image.asset(
                          _voiceImage,
                          width: 100,
                          height: 100,
                          package: 'flutter_sound_suite',
                        ),
                      ),
                      Container(
                        child: Text(
                          _overLayHint,
                          style: TextStyle(
                            fontStyle: FontStyle.normal,
                            color: Colors.white,
                            fontSize: 14,
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
          ),
        );
      });
      Overlay.of(context)!.insert(_overlayEntry!);
    }
  }

  updateStatus() {
    setState(() {
      _isCancel = _startY - _offset > 100 ? true : false;
      if (_isCancel) {
        _buttonHint = '松开 取消';
        _overLayHint = '下滑 继续';
        if (_recorder.isRecording) {
          _recorder.pauseRecorder();
        }
      } else {
        _buttonHint = '松开 结束';
        _overLayHint = '上滑 暂停';
        if (_recorder.isPaused) {
          _recorder.resumeRecorder();
        }
      }
    });
  }
}
